docs/hugo/content/design/ADR-2024-10-OneOf-Resources/_index.md

```diff
@@ -0,0 +1,459 @@
+
+## Option 4: Special case root one-of types
+
+Preserve the existing rules for one-of types, but special case the root level of a resource spec to permit additional properties (including both `AzureName` and `OperatorSpec`) be specified alongside the one-of properties.
+
+``` mermaid
+classDiagram
+
+class Role {
+    AzureName string
+    OperatorSpec RoleOperatorSpec
+    Student StudentProperties
+    Tutor TutorProperties
+    Teacher TeacherProperties
+    Marker MarkerProperties
+}
+
+class StudentProperties {
+    Class string
+    Grade string
+    StartDate date
+    FinishDate optional<date>
+}
+
+
+class TutorProperties {
+    Class string
+    Timeslot string
+    StartDate date
+    FinishDate optional<date>
+}
+
+class TeacherProperties {
+    Subject string
+    Tenure bool
+    StartDate date
+    FinishDate optional<date>
+}
+
+class MarkerProperties {
+    Exam string
+    Strictness string
+    StartDate date
+    FinishDate optional<date>
+}
+
+Role *-- StudentProperties : Student
+Role *-- TutorProperties : Tutor
+Role *-- TeacherProperties : Teacher
+Role *-- MarkerProperties : Marker
+```
+
+
+### Pros
+
+* Limits the scope of impact of the change.
+
+### Cons
+
+* Need to change the generation of our JSON marshalling code to handle this special case.
+* Special casing is always a bit of a code smell.
+
+
+## Option 5: Push the one-of down a level
+
+Push all of the _one-of_ properties down from the root into a nested object, and then have the root object contain `AzureName` and other top level properties. We'd call this `Properties` to be consistent with normal ARM conventions.
+
+``` mermaid
+classDiagram
+
+class Role {
+    AzureName string
+    OperatorSpec RoleOperatorSpec
+    Properties RoleProperties
+}
+
+class RoleProperties {
+    Student StudentProperties
+    Tutor TutorProperties
+    Teacher TeacherProperties
+    Marker MarkerProperties
+}
+
+class StudentProperties {
+    Class string
+    Grade string
+    StartDate date
+    FinishDate optional<date>
+}
+
+
+class TutorProperties {
+    Class string
+    Timeslot string
+    StartDate date
+    FinishDate optional<date>
+}
+
+class TeacherProperties {
+    Subject string
+    Tenure bool
+    StartDate date
+    FinishDate optional<date>
+}
+
+class MarkerProperties {
+    Exam string
+    Strictness string
+    StartDate date
+    FinishDate optional<date>
+}
+
+Role *-- RoleProperties : Properties
+RoleProperties *-- StudentProperties : Student
+RoleProperties *-- TutorProperties : Tutor
+RoleProperties *-- TeacherProperties : Teacher
+RoleProperties *-- MarkerProperties : Marker
+```
+
+### Pros
+
+* Easily to implement using a new pipeline stage in the code generator (but: ease of implementation is only one factor of many).
+
+### Cons
+
+* Adds yet another level of nesting to affected resources when we've done a lot of prior work to flatten things out.
+* Potentially confusing to users expecting the structure to more or less match the ARM API 
```

Comment:
I don't know (w.r.t. OneOf) how much of an issue this really is, because we're already different than the raw ARM API, as are some of the SDKs because most languages don't support true OneOf, at least not in a user-friendly way.

I think the biggest con here is just the extra level of nesting and what confusion(s) that might bring.